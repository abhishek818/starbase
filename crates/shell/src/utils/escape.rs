use std::fmt::Write;

// referenced from 'https://github.com/solidsnack/shell-escape/blob/master/Data/ByteString/ShellEscape/Bash.hs'

#[derive(Debug, PartialEq)]
enum EscapingMode {
    ANSIHex,
    ANSIBackslash,
    Literal,
    Quoted,
}

fn classify(c: char) -> EscapingMode {
    match c {
        '\u{0}'..='\u{6}' => EscapingMode::ANSIHex,    // \x00-\x06: ANSIHex
        '\u{9}' => EscapingMode::ANSIBackslash,        // \t: ANSIBackslash
        '\u{a}' => EscapingMode::ANSIBackslash,        // \n: ANSIBackslash
        '\u{d}' => EscapingMode::ANSIBackslash,        // \r: ANSIBackslash
        '\u{1f}' => EscapingMode::ANSIHex,             // \x1f: ANSIHex
        '\u{20}'..='\u{26}' => EscapingMode::Quoted,   // ' ' - '&': Quoted
        '\'' => EscapingMode::ANSIBackslash,           // '\'': ANSIBackslash
        '\u{2b}' => EscapingMode::Quoted,              // '+': Quoted
        '\u{30}'..='\u{39}' => EscapingMode::Literal,  // '0' - '9': Literal
        '\u{3f}' => EscapingMode::Quoted,              // '?': Quoted
        '\u{41}'..='\u{5a}' => EscapingMode::Literal,  // 'A' - 'Z': Literal
        '[' => EscapingMode::Quoted,                   // '[': Quoted
        '\\' => EscapingMode::ANSIBackslash,           // '\\': ANSIBackslash
        '_' => EscapingMode::Literal,                  // '_': Literal
        '\u{5d}' => EscapingMode::Quoted,              // ']': Quoted
        '\u{60}' => EscapingMode::Quoted,              // '`': Quoted
        '\u{61}'..='\u{7a}' => EscapingMode::Literal,  // 'a' - 'z': Literal
        '\u{7e}' => EscapingMode::Quoted,              // '~': Quoted
        '\u{7f}' => EscapingMode::ANSIHex,             // \x7f: ANSIHex
        _ => EscapingMode::ANSIHex,                    // Default: ANSIHex
    }
}

fn hexify(c: char) -> String {
    format!("\\x{:02X}", c as u32)
}

fn backslashify(c: char) -> String {
    format!("\\{}", c)
}

pub fn escapeBash(input: &str) -> String {
    let mut output = String::new();
    let mut escape_needed = false;

    for c in input.chars() {
        match classify(c) {
            EscapingMode::ANSIHex => {
                write!(output, "{}", hexify(c)).unwrap();
                escape_needed = true;
            }
            EscapingMode::ANSIBackslash => {
                write!(output, "{}", backslashify(c)).unwrap();
                escape_needed = true;
            }
            EscapingMode::Quoted => {
                write!(output, "{}", c).unwrap();
                escape_needed = true;
            }
            EscapingMode::Literal => {
                write!(output, "{}", c).unwrap();
            }
        }
    }

    if escape_needed {
        format!("$'{}'", output)
    } else {
        output
    }
}
