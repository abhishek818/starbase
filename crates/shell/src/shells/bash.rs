use super::Shell;
use crate::utils::escape::escapeBash;
use crate::hooks::Hook;
use std::fmt;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug)]
pub struct Bash;

impl Bash {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self
    }
}

// https://www.baeldung.com/linux/bashrc-vs-bash-profile-vs-profile
impl Shell for Bash {
    fn format_env_set(&self, key: &str, value: &str) -> String {
        // format!("export {}={};", self.quote(key), self.quote(value))
        format!("export {}={};", escapeBash(key), escapeBash(value))
    }

    fn format_env_unset(&self, key: &str) -> String {
        // format!("unset {key};", key=self.quote(key))
        format!("unset {key};", key=escapeBash(key))
    }    

    fn format_path_set(&self, paths: &[String]) -> String {
        // let escaped_paths: Vec<String> = paths.iter().map(|p| self.quote(p)).collect();
        // format!(r#"export PATH="{}:$PATH";"#, escaped_paths.join(":"))
        let escaped_paths: Vec<String> = paths.iter().map(|p| escapeBash(p)).collect();
        format!(r#"export PATH="{}:$PATH";"#, escaped_paths.join(":"))
    }

    fn format_hook(&self, hook: Hook) -> Result<String, crate::ShellError> {
        Ok(hook.render_template(
            self,
            r#"
_{prefix}_hook() {
  local previous_exit_status=$?;
  trap -- '' SIGINT;
{export_env}
{export_path}
  trap - SIGINT;
  return $previous_exit_status;
};

if [[ ";${PROMPT_COMMAND[*]:-};" != *";_{prefix}_hook;"* ]]; then
  if [[ "$(declare -p PROMPT_COMMAND 2>&1)" == "declare -a"* ]]; then
    PROMPT_COMMAND=(_{prefix}_hook "${PROMPT_COMMAND[@]}")
  else
    PROMPT_COMMAND="_{prefix}_hook${PROMPT_COMMAND:+;$PROMPT_COMMAND}"
  fi
fi
"#,
            "  ",
        ))
    }

    fn get_config_path(&self, home_dir: &Path) -> PathBuf {
        home_dir.join(".bash_profile")
    }

    fn get_env_path(&self, home_dir: &Path) -> PathBuf {
        home_dir.join(".bash_profile")
    }

    fn get_profile_paths(&self, home_dir: &Path) -> Vec<PathBuf> {
        vec![
            home_dir.join(".bash_profile"),
            home_dir.join(".bashrc"),
            home_dir.join(".profile"),
        ]
    }

    // TODO: commented out to get some basic tests work first
    // fn quote(&self, value: &str) -> String {
    //     escapeBash(value)
    // }
}

impl fmt::Display for Bash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "bash")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    #[test]
    fn formats_env_var() {
        assert_eq!(
            Bash.format_env_set("PROTO_HOME", "$HOME/.proto"),
            r#"export PROTO_HOME="$HOME/.proto";"#
        );
    }

    #[test]
    fn formats_path() {
        assert_eq!(
            Bash.format_path_set(&["$PROTO_HOME/shims".into(), "$PROTO_HOME/bin".into()]),
            r#"export PATH="$PROTO_HOME/shims:$PROTO_HOME/bin:$PATH";"#
        );
    }

    #[test]
    fn formats_cd_hook() {
        let hook = Hook::OnChangeDir {
            env: vec![
                ("PROTO_HOME".into(), Some("$HOME/.proto".into())),
                ("PROTO_ROOT".into(), None),
            ],
            paths: vec!["$PROTO_HOME/shims".into(), "$PROTO_HOME/bin".into()],
            prefix: "starbase".into(),
        };

        assert_snapshot!(Bash.format_hook(hook).unwrap());
    }

    #[test]
    fn test_escape_plain_string() {
        // let bash_shell = Bash::new();
        assert_snapshot!(escapeBash("foobar"));
    }

    #[test]
    fn test_escape_string_with_spaces() {
        // let bash_shell = Bash::new();
        assert_snapshot!(escapeBash("foo bar"));
    }

    #[test]
    fn test_escape_string_with_single_quotes() {
        // let bash_shell = Bash::new();
        assert_snapshot!(escapeBash("don't"));
    }

    #[test]
    fn test_escape_string_with_special_characters() {
        // let bash_shell = Bash::new();
        assert_snapshot!(escapeBash("$100@*&"));
    }

    #[test]
    fn test_escape_string_with_backslashes() {
        // let bash_shell = Bash::new();
        assert_snapshot!(escapeBash("a\\b"));
    }

    #[test]
    fn test_escape_empty_string() {
        // let bash_shell = Bash::new();
        assert_snapshot!(escapeBash(""));
    }
}
