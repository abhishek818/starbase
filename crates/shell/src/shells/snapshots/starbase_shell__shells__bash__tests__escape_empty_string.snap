---
source: crates/shell/src/shells/bash.rs
assertion_line: 156
expression: "escapeBash(\"\")"
---

